provisioner/aws-manager.js

```diff
@@ -0,0 +1,696 @@
+'use strict';
+
+var Promise = require('promise');
+var lodash = require('lodash');
+var util = require('util');
+var assert = require('assert');
+var debug = require('debug')('aws-provisioner:aws-manager');
+
+
+/**
+ * AWS EC2 state at a specific moment in time
+ */
+function AwsManager(ec2, keyPrefix, pubKey) {
+  assert(ec2);
+  assert(keyPrefix);
+  assert(pubKey);
+  this.ec2 = ec2;
+  this.keyPrefix = keyPrefix;
+  this.pubKey = pubKey;
+  this.__apiState = {};
+  this.__internalState = {};
+  this.__knownKeyPairs = [];
+}
+
+module.exports = AwsManager;
+
+
+/**
+ * Update the state from the AWS API and return a promise
+ * with no resolution value when completed.
+ */
+AwsManager.prototype.update = function() {
+  var that = this;
+
+  var p = Promise.all([
+    that.ec2.describeInstances({
+      Filters: [{
+        Name: 'key-name',
+        Values: [that.keyPrefix + '*']
+      },{
+        Name: 'instance-state-name',
+        Values: ['running', 'pending']
+      }
+    ]}),
```

Comment:
I hate to point this out, but `describeInstances` may return an `NextToken` property, as well as it make take this property as input. Basically the API requires paging.

The old provisioner doesn't do this, and I suspect that the option is only there because AWS has a really old API.
But the property is there, and EC2 APIs are notoriously slow, so we better implement it... And no you can't really implement this with multi-region-aws-sdk. You need to do per region paging...